### PR TITLE
[imp] Implement Backend Language Switcher

### PR DESCRIPTION
**Pull request by infograf768**
This PR adds a new Toolbar Button to let admin switch the Backend Language to any language.
It replaces https://github.com/joomla/joomla-cms/pull/9119 which was only letting switch to the Default Admin Language.

The purpose of the PR is to facilitate the work of devs, translators and admins in general.

After PR, one will get:
screen shot 2016-02-15 at 09 14 16

@richard67 @andrepereiradasilva @Kubik-Rubik 
Thanks for testing.

### CODE REVIEW

File: administrator/components/com_languages/controllers/installed.php [hunk 1]
```diff
@@ -43,4 +43,32 @@ public function setDefault()
...
 		$this->setredirect('index.php?option=com_languages&view=installed&client=' . $clientId, $msg, $type);
 	}
+
+	/**
+	 * Task to switch the administrator language.
+	 *
+	 * @return  void
+	 */
+	public function switchAdminLanguage()
+	{
```
Comment (dgt41) [inline administrator/components/com_languages/controllers/installed.php]: If you allow a param in this function ($languageId) then we can use it in a module lang switcher.
Also a check 
```php
if (isset($languageId))
{
  JFactory::getApplication()->setUserState('application.lang', $languageId);
}
else
{
$cid   = $this->input->get('cid', ''); 
...

}
```

**Comment by andrepereiradasilva** [comment]
@infograf768 for usability how about reordering the buttons, first the ones related with options selected, them the general ones (not related to the options selected).

Also what to do you think of instead "Switch Backend Language" having "Switch to Language". Since the button only appear on the admin languages, IMHO probably makes more sense.

In other words:

instead of: Default | Install Languages | Switch Backend Language

this: Default | Switch to Language | Install Languages

File: administrator/components/com_languages/controllers/installed.php (hunk 1 above)
Comment (infograf768) [inline administrator/components/com_languages/controllers/installed.php]: You mean we "could" use it, right?
Do you want to add one to core or would that just be to let a 3rd party create one?

Also, I would rather, if @joomla/cms-maintainers agree with your proposal, use $languageTag as it is not an id which is for $whatever in
`JFactory::getApplication()->setUserState('application.lang', $whatever)'

but the language tag, i.e fr-FR, en-GB, etc.

**Comment by infograf768** [comment]
@andrepereiradasilva 
Ok for the order. Will wait all comments to update.
Asking @brianteeman for the string change.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on dbd69203887e8adb49e0933623fa29e842d24d8d<br /><br />Tested with success. Works as discussed in #9119 at the end.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9126">issues.joomla.org/joomla-cms/9126</a>.</sub>

File: administrator/components/com_languages/controllers/installed.php (hunk 1 above)
Comment (dgt41) [inline administrator/components/com_languages/controllers/installed.php]: Yup!
I find this an essential addition for backend users with multilingual sites. Right now I need to install 3PD extension for this really basic functionality.
Figuring out the variable by its name is hard but yes $languageTag seems more appropriate!

**Comment by MATsxm** [comment]
I have tested this item :white_check_mark: successfully on dbd69203887e8adb49e0933623fa29e842d24d8d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9126">issues.joomla.org/joomla-cms/9126</a>.</sub>

**Comment by MATsxm** [comment]
At this point - works as described.

>> NOTE: the btn appears even if we have only one language istalled on backend
screen shot 2016-02-15 at 06 08 50<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9126">issues.joomla.org/joomla-cms/9126</a>.</sub>

**Comment by brianteeman** [comment]
For consistency I think that this should be referred to as Administrator Language and not Backend Language<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9126">issues.joomla.org/joomla-cms/9126</a>.</sub>

**Comment by infograf768** [comment]
I have modified the language strings as well as the order of the buttons.

This gives:
screen shot 2016-02-16 at 08 00 20

@MATsxm  The button will behave the same way as the "Default" button which also displays when only one language is installed.

@andrepereiradasilva  I guess that Switch Language does not need the "to". 

Please test again

@dgt41 
I prefer to let you make a specific PR for your request as I would not know where to stop the "else".

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 295bf8c1e8531384486ef076052d593c4042e2f4<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9126">issues.joomla.org/joomla-cms/9126</a>.</sub>

**Comment by MATsxm** [comment]
I have tested this item :white_check_mark: successfully on 295bf8c1e8531384486ef076052d593c4042e2f4<br /><br />#9126 works as expected - Thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9126">issues.joomla.org/joomla-cms/9126</a>.</sub>

**Comment by infograf768** [comment]
rtc. thanks for feedback and testing.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9126">issues.joomla.org/joomla-cms/9126</a>.</sub>

**Comment by infograf768** [comment]
Folks, I will ask you to test again, sorry.
@brianteeman suggested that I change the success message to display the language Name and not the language Tag.
I think it is indeed much better.

One will now get:
screen shot 2016-02-17 at 12 24 55

@richard67 @MATsxm @andrepereiradasilva @brianteeman

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 49595b0a6692a2e51c041489060ef0891b4af47b<br /><br />Just tested, works as before plus language full name instead of tag is used in the success message.

@infograf768 What last screenshot does not show (because I think you switched from English to English) is that the success message is in the language from before switching, as your first screenshot in the test description shows (switched from English to Italian: Title "Messagio" in Italian but message content in English. I guess we cannot change this in an easy way, otherwise you would have done it, right? Or is it by purpose?

I also assume that there is no other easy way to get the language full name so you have to parse the XML file, right?

Or could it maybe be obtained from the language lists's 1st column, e.g. with jQuery?

Anyway, for me it is good enough as it is now, and I hope it will go into 3.5.0 Beta 3 or RC.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9126">issues.joomla.org/joomla-cms/9126</a>.</sub>

**Comment by MATsxm** [comment]
I have tested this item :white_check_mark: successfully on 49595b0a6692a2e51c041489060ef0891b4af47b<br /><br />Thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9126">issues.joomla.org/joomla-cms/9126</a>.</sub>

**Comment by infograf768** [comment]
>as your first screenshot in the test description shows (switched from English to Italian: Title "Messagio" in Italian but message content in English. I guess we cannot change this in an easy way, otherwise you would have done it, right?

Same issue when the user is set to Default Language and another language is set to default. The content of the message is picked from the former language. No idea how to solve this. If it is possible this should be another PR imho.

I did parse the xx-XX.xml file indeed. Easy enough. code is
```
		$file = JPATH_ADMINISTRATOR . '/language/' . $cid . '/' . $cid . '.xml';
		$info = JInstaller::parseXMLInstallFile($file);
		$languageName = $info['name'];

		if ($model->switchAdminLanguage($cid))
		{
			$msg = JText::sprintf('COM_LANGUAGES_MSG_SWITCH_ADMIN_LANGUAGE_SUCCESS', $languageName);
			$type = 'message';
		}
		else
```

**Comment by richard67** [comment]
@infograf768 Thanks for the info, I just was curious. For me it is fine now as it is.

**Comment by andrepereiradasilva** [comment]
i think you could also use `JLanguage::parseXMLLanguageFile($path);` don't really know the difference.

**Comment by brianteeman** [comment]
@richard67 @infograf768 isnt the message always in english right now because it is a new language string that doesnt exist in the other languages so en-gb is always used as a fallback <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9126">issues.joomla.org/joomla-cms/9126</a>.</sub>

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 49595b0a6692a2e51c041489060ef0891b4af47b<br /><br />Works perfectly for me - thanks @infograf768 <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9126">issues.joomla.org/joomla-cms/9126</a>.</sub>

**Comment by richard67** [comment]
@brianteeman Ahhhh, sure, you are right. Silly me!

**Comment by infograf768** [comment]
@andrepereiradasilva 

>i think you could also use JLanguage::parseXMLLanguageFile($path); don't really know the difference.

It is deprecated.
https://github.com/joomla/joomla-cms/blob/staging/libraries/cms/application/helper.php#L229-L231

**Comment by andrepereiradasilva** [comment]
not that one, this one
https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/language/language.php#L1365

**Comment by gwsdesk** [comment]
Works as intended for me. Thanks JM

**Comment by richard67** [comment]
@infograf768 Thanks a lot for this great new feature. It helps a lot with testing (multi) language stuff in backend. I think it is worth to be mentioned in the new features of the release notes with an own head line.

**Comment by infograf768** [comment]
BTW:
>isnt the message always in english right now because it is a new language string that doesnt exist in the other languages so en-gb is always used as a fallback 

Nope. As I stated above https://github.com/joomla/joomla-cms/pull/9126#issuecomment-185170363 the message is in the former language. It is the same as Default change as you can test by setting an admin language to Default when logged with a user set to use Default language.

**Comment by infograf768** [comment]
I have a solution for that. Will make a new PR for both Default and Switch.

**Pull request merged**